### PR TITLE
Allow target function to output columns already present in the data set

### PR DESCRIPTION
**Pull request by nick863**
# Description

Currently target function will fail if its output column will be present in the input data set. In this PR we are fixing it. See work item #3170343.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [x] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by nagkumar91 — APPROVED** [review]

**Review by luigiw — APPROVED** [review]

File: src/promptflow-evals/promptflow/evals/evaluate/_evaluate.py [hunk 1]
```diff
@@ -163,17 +171,34 @@ def _apply_column_mapping(source_df: pd.DataFrame, mapping_config: dict, inplace
...
                 pattern = match.group(1)
                 if pattern.startswith(pattern_prefix):
-                    map_from_key = pattern.split(pattern_prefix)[1]
-                    column_mapping[map_from_key] = map_to_key
-
-        result_df = source_df.rename(columns=column_mapping, inplace=inplace)
+                    column_mapping[pattern[len(pattern_prefix):]] = map_to_key
+                elif pattern.startswith(run_outputs_prefix):
+                    map_from_key = pattern[len(run_outputs_prefix):]
+                    col_outputs = f'outputs.{map_from_key}'
```
[the remark this answers is not in the document]
Comment (nick863) [reply src/promptflow-evals/promptflow/evals/evaluate/_evaluate.py:186]: The main issue here is what would happen if user will have something like 
"new_col": ${run.outputs.col1}, "new_col": "another_col", it adds a litle bit of complexity as in this case we should not drop column.

File: src/promptflow-evals/promptflow/evals/evaluate/_evaluate.py [hunk 2]
```diff
@@ -256,60 +308,70 @@ def evaluate(
...
+            for col in target_generated_columns:
+                # If user defined mapping differently, do not change it.
+                # If it was mapped to target, we have already changed it
+                # in _process_evaluator_config
+                run_output = f'${{run.outputs.{col}}}'
+                # We will add our mapping only if
+                # customer did not mapped target output.
+                if col not in mapping and run_output not in mapped_to_values:
+                    evaluator_config[evaluator_name][col] = run_output
+
```
[the remark this answers is not in the document]
Comment (nick863) [reply src/promptflow-evals/promptflow/evals/evaluate/_evaluate.py:329]: In this case we are making sure, we map all the outputs of target function to the apropriate columns in the evaluators input, however, we should not do that if customer have mapped this output to something else.

File: src/promptflow-evals/promptflow/evals/evaluate/_evaluate.py [hunk 3]
```diff
@@ -139,19 +137,29 @@ def _apply_target_to_data(
...
+    # Sometimes user data may contain column named the same as the one generated by target.
+    # In this case we will not rename the column.
+    generated_columns = set(rename_dict.values())
+    tgt_out_and_input = set()
+    for col in initial_data.columns:
+        if col in generated_columns:
+            tgt_out = f'{prefix}{col}'
+            del rename_dict[tgt_out]
+            tgt_out_and_input.add(tgt_out)
+            tgt_out_and_input.add(col)
```
[the remark this answers is not in the document]
Comment (nick863) [reply src/promptflow-evals/promptflow/evals/evaluate/_evaluate.py:149]: Fixed.

File: src/promptflow-evals/promptflow/evals/evaluate/_evaluate.py [hunk 4]
```diff
@@ -199,12 +224,39 @@ def _process_evaluator_config(evaluator_config: Dict[str, Dict[str, str]]):
                             "Ensure only ${target.} and ${data.} are used."
                         )
 
-                    # Replace ${target.} with ${data.}
-                    processed_config[evaluator][map_to_key] = map_value.replace("${target.", "${data.")
+                    # Replace ${target.} with ${run.}
```
[the remark this answers is not in the document]
Comment (nick863) [reply src/promptflow-evals/promptflow/evals/evaluate/_evaluate.py:227]: Fixed

File: src/promptflow-evals/promptflow/evals/evaluate/_evaluate.py [hunk 5]
```diff
@@ -139,19 +137,29 @@ def _apply_target_to_data(
...
     target_output.drop(drop_columns, inplace=True, axis=1)
     # Remove outputs. prefix
     target_output.rename(columns=rename_dict, inplace=True)
     # Concatenate output to input
     target_output = pd.concat([target_output, initial_data], axis=1)
-    return target_output, set(rename_dict.values()), run
+    return target_output, generated_columns, run
 
 
 def _apply_column_mapping(source_df: pd.DataFrame, mapping_config: dict, inplace: bool = False):
```
[the remark this answers is not in the document]
Comment (nick863) [reply src/promptflow-evals/promptflow/evals/evaluate/_evaluate.py:165]: Added docstring.

**Pull request merged**